gnocchi/rest/api.py

```diff
@@ -1890,6 +1899,106 @@ class BatchController(object):
     resources = ResourcesBatchController()
 
 
+class PrometheusWriteController(rest.RestController):
+
+    PROMETHEUS_RESOURCE_TYPE = {
+        "instance": {"type": "string",
+                     "min_length": 1,
+                     "max_length": 512,
+                     "required": True},
+        "job": {"type": "string",
+                "min_length": 1,
+                "max_length": 512,
+                "required": True}
+    }
+
+    # Retry with exponential backoff for up to 1 minute
+    @classmethod
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=0.5, max=60),
+        retry=tenacity.retry_if_exception_type(
+            (indexer.NoSuchResource, indexer.ResourceAlreadyExists,
+             indexer.NamedMetricAlreadyExists)))
+    def get_or_create_resource_and_metrics(cls, creator, rid,
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
```

Comment:
If `rid` already exists with type != `prometheus` then r will be None, but `create_resource` on L1956 will raise `indexer.ResourceAlreadyExists`, which will make this retry, and loop forever.

Reply:
I have fixed it, but honestly I will be a uuid collision, so...